pkg/promtail/client/client.go

```diff
@@ -155,8 +158,18 @@ func (c *Client) sendBatch(batch map[model.Fingerprint]*logproto.Stream) {
 	var status int
 	for backoff.Ongoing() {
 		start := time.Now()
-		status, err = c.send(ctx, buf)
+		timedCtx, ctxCancel := context.WithTimeout(ctx, c.cfg.Timeout)
+		status, err = c.send(timedCtx, buf)
 		requestDuration.WithLabelValues(strconv.Itoa(status)).Observe(time.Since(start).Seconds())
+		ctxErr := timedCtx.Err()
+		if ctxErr != nil {
+			switch ctxErr {
+			case context.DeadlineExceeded:
+				err = errors.Wrap(err, "Timeout waiting for server")
+			}
+		}
```

Comment:
I'm not sure you need to check the context.Err here - the `client.Do()` in `c.send` will return what ever error comes from the context: https://golang.org/src/net/http/transport.go?s=3607:10104#L451

Reply:
the main reason for all this mess was to clarify the log message, without this you get a message that looks like 
```
level=warn msg="error sending batch, will retry" status=-1 error="Post http://10.10.10.10:3100/api/prom/push: context deadline exceeded"
```
So I added the check for the specific error to be able to add the "Timeout waiting for server" to the message.

I'll defer to you if you think the built in message is descriptive enough?

Reply:
`context deadline exceeded` is becoming a pretty common synonym for `timeout`, so yeah, I'd got for less code.